### DISCUSSION TITLE
Prisma 2 vs Mysql driver - Simple performance test

### DISCUSSION THREAD

**Discussion by m-arrieta-r** [prisma/prisma#5819] [Q&A] [answered]
I did the following test:
https://gist.github.com/marnutux/755a50c6569d9d5947fb05aa907cd2d8

According to that test Prisma is almost 10 times slower than mysql driver. I understand that the team is focused on development productivity, and it is obvious that it is going to be slower, but it sounds to me a lot.

**Comment by nikolasburk** [comment, marked as answer]
Hey @marnutux, thank you so much for sharing this! I fully agree that this performance difference is quite a bit and we definitely want to improve the situation! (Just one note: While the difference of 10x is quite notable, I think the comparison would be more fair if it actually compared Prisma against another ORM rather than the raw SQL driver which of course is expected to be faster. Nonetheless, I agree that 10x is still too much here.)

We've generally been following a "make it work, make it right, make it fast" philosophy for building Prisma and are just arriving at the "make it fast" stage where we start to improve performance.

We've recently started an internal benchmarking effort (led by @jasonkuhrt) where we compare Prisma Client's performance against other ORMs.  We'll use this to identify and eliminate bottlenecks in Prisma Client queries.

Also, due to some legacy architectural decisions, the Node.js language layer of Prisma Client currently communicates via HTTP with the underlying query engine that generates the SQL queries and that's implemented in Rust. This obviously creates quite some overhead. The good news here is that [we are working towards replacing this interface](https://github.com/prisma/prisma/issues/4911) using the [N-API](https://napi.rs/) which will allow to _directly_ call into the Rust code of the query engine without the overhead of HTTP. We are getting very close to having this implemented (it's coming _most likely_ within the next 4-6 weeks), you can follow the development on the roadmap [here](https://www.notion.so/prismaio/N-API-Support-05887bbd99f14f23936e67759ece35f1).

Hope this helps, please let me know if you have any further questions! 🙌

**Reply by nikolasburk** [reply]
Ah, one more note @marnutux:

When measuring Prisma Client performance, you should also explicitly call `await prisma.$connect()` _before_ you send your first query. Otherwise, due to the lazy connection behaviour of Prisma Client, your measurement will include the time that it takes to establish a DB connection! 🙏

**Reply by michael-land** [reply]
Will replacing n-api address unnecessary select like this one ? #4246

**Reply by nikolasburk** [reply]
I don't think so, @xiaoyu-tamu. This is only about the interface between Prisma Client's Node.js language layer with the query engine, not about the SQL queries that are generated by the query engine.

**Reply by michael-land** [reply]
Ah, Thanks. I still need to wait until this get optimized.

**Reply by chinanderm** [reply]
@nikolasburk Thank you for this fabulous insight Nikolas! I'm really looking forward to the performance updates.

**Reply by lilletech** [reply]
Is the problem still present ? 
I want to adopt Prisma for a project but this discussion is scary!

**Reply by nikolasburk** [reply]
@lilletech We haven't any performance benchmarks ourselves yet but with everything that I've already mentioned above and the [GA release of the Node-API](https://www.prisma.io/blog/wnip-q3-hpk7pyth8v#node-api-is-generally-available), I'm pretty sure that this is a non-problem :) 

Let me know if you have any further questions @lilletech, always happy to help 🙌

**Reply by yf-hk** [reply]
Seems like in Prisma 3.0 N-API is the default mode? @marnutux have you tested again?

**Reply by ghost** [reply]
This reply is so perfect in every regard: satisfied user, verified they are working on extending and improving performance while still focusing on their original goal. The developer team is clearly very good. I'm switching my projects from mongoose to Prisma ASAP!

**Comment by lassegit** [comment]
Is data about RAM usage comparisons?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
